text/0000-externally-implementable-functions.md

```diff
@@ -0,0 +1,140 @@
+- Feature Name: `extern_impl_fn`
+- Start Date: 2024-05-10
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+A mechanism for defining a function whose implementation can be defined (or overridden) in another crate.
+
+Example 1:
+
+```rust
+// core::panic:
+
+extern impl fn panic_handler(_: &PanicInfo) -> !;
+
+// user:
+
+impl fn core::panic::panic_handler(_: &PanicInfo) -> ! {
+    loop {}
+}
+```
+
+Example 2:
+
+```rust
+// log crate:
+
+extern impl fn logger() -> Logger {
+    Logger::default()
+}
+
+// user:
+
+impl fn log::logger() -> Logger {
+    Logger::to_stdout().with_colors()
+}
+```
+
+# Motivation
+
+We have several items in the standard library that are overridable/definable by the user crate.
+For example, the (no_std) `panic_handler`, the global allocator for `alloc`, and so on.
+
+Each of those is a special lang item with its own special handling.
+Having a general mechanism simplifies the language and makes this functionality available for other crates, and potentially for more use cases in core/alloc/std.
+
+# Explanation
+
+A function can be defined as "externally implementable" using `extern impl` as follows:
+
+```rust
+// In crate `x`:
+
+// Without a body:
+extern impl fn a();
+
+// With a body:
+extern impl fn b() {
+    println!("default impl");
+}
+```
+
+Another crate can then provide (or override) the implementation of these functions using `impl fn` syntax (using their path) as follows:
+
+```rust
+// In another crate:
+
+impl fn x::a() {
+    println!("my implementation of a");
+}
+
+impl fn x::b() {
+    println!("my implementation of b");
+}
+```
+
+# Details
+
+## Signature
+
+It is an error to have a different signature for the `impl fn` item
+
+## No impl
+
+It is an error to have no `impl fn` item (in any crate) for an `extern impl fn` item without a body.
+
+## Duplicates
+
+It is an error to have multiple `impl fn` items (across all crates) for the same `extern impl fn` item.
+
+## Visibility
+
+`extern impl fn` items can have a visibility specifier (like `pub`), which determines who can *call* the function (or create pointers to it, etc.).
+
+*Implementing* the function can be done by any crate that can name the item.
```

Comment:
Could you clarify whether this is intended to also tie into visibility? For instance, a `pub(crate) extern impl fn` can only have an implementation provided by the crate, right?

Reply:
The idea is that if you do:

```rust
pub mod a {
    pub(crate) extern impl fn x();
}
```

The, other crates can provide an implementation (because `a` is `pub`, allowing them to name `a::x`), but they cannot call it (because the function is not public itself).

Reply:
@m-ou-se That seems potentially confusing. Is there some way to set the visibility of being able to *implement* it? Is there value in being able to do so?

Reply:
No, because the point of this feature is allowing other crates to implement it. So unless you want to propose a kind of visibility that includes some crates but not others, that just implies full public.

Reply:
So if the `extern impl fn` is in a private module, it is impossible to implement?

Reply:
@m-ou-se 

> The idea is that if you do:
> 
> ```rust
> pub mod a {
>     pub(crate) extern impl fn x();
> }
> ```
>
> The, other crates can provide an implementation (because `a` is `pub`, allowing them to name `a::x`), but they cannot call it (because the function is not public itself).

Yeah this is very confusing. Could you make it respect the normal privacy rule and reuse e.g. #3323 to explicitly deny call/ref permission from dependencies?

```rust
pub mod a {
    pub restrict_use(crate) extern impl fn x();
}
```

Reply:
> So if the extern impl fn is in a private module, it is impossible to implement?

It could be re-exported, but that reexport then presumably allows *both* calling and implementing the item. The magic behavior where these two permissions are different can only be obtain via the original definition of the item, IIUC.